Full domain chain snap sync (v2).

shamil-gadelshin opened the pull request:
This PR introduces snap sync for the domain chain (https://github.com/autonomys/subspace/issues/3026).

It supersedes the previous [PR](https://github.com/autonomys/subspace/pull/3115) with squashed review commits into the original code to enable clean review (requested by @nazar-pc [here](https://github.com/autonomys/subspace/pull/3115#issuecomment-2437558737))

The main changes from the original PR:
- the last confirmed domain block execution receipt protocol was moved to domain network from consensus network
- the last confirmed domain block execution receipt sync algorithm was changed to support possible change due to network delays
- `SubspaceSyncOracle` was modified to support domain snap sync
- `SnapSyncOrchestrator` was simplified significantly
- consensus and domain chains synchronization was refactored from `SnapSyncOrchestrator` to block import notifications

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/autonomys/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
Add domain snap sync algorithm.

shamil-gadelshin committed:
Remove last confirmed domain block receipt handlers.

shamil-gadelshin committed:
Update test dependencies (domain snap-sync integration)

shamil-gadelshin committed:
Update domain service (domain snap-sync integration)

shamil-gadelshin committed:
Update subspace-node (domain snap-sync integration)

shamil-gadelshin committed:
Update subspace-service (domain snap-sync integration)

shamil-gadelshin committed:
Update MMR-sync

shamil-gadelshin committed:
Introduce snap sync orchestrator.

shamil-gadelshin committed:
Update domain operator (domain snap-sync integration)

teor2345 reviewed (commented):
Can you merge the `main` branch into this branch and resolve conflicts?
Otherwise we’re going to have to review this PR again before it merges.

(Anyone who wants to compare with the original PR, or review without the merge, can do that up to the last current commit, which is 21ee3e04cc5e717f83dfedee173126eaca808a0f.)

shamil-gadelshin committed:
Modify SubspaceSyncOracle to add domain snap sync.

shamil-gadelshin committed:
Modify domain networking stack.

- add custom `build-network` method
- add last domain block execution receipt protocol (moved from consensus chain)

shamil-gadelshin committed:
Update consensus snap-sync

nazar-pc commented on crates/sc-consensus-subspace/src/slot_worker.rs:
```diff
@@ -106,6 +107,11 @@ where
         // (default state), it also accounts for DSN sync
         (!self.force_authoring && self.inner.is_major_syncing())
             || self.pause_sync.load(Ordering::Acquire)
+            || self
```
This oracle is about consensus chain sync. It shouldn't know or care about domain sync state.

shamil-gadelshin replied:
Hmm, I was under impression that you suggested the very same idea as Ning which I implemented. What did I miss?

nazar-pc replied:
My understanding was that this was needed in domains, in which case you create a domain sync oracle with necessary custom logic there, but consensus sync oracle doesn't need to be modified

teor2345 commented on domains/service/src/network.rs:
```diff
@@ -0,0 +1,209 @@
+            client.clone(),
+            num_peer_hint,
+        );
+        spawn_handle.spawn(
```
Why aren’t any of the tasks in this function essential?

shamil-gadelshin replied:
It's a copy of `build_network()` code from Substrate. The logic seems to be that any network protocol task failure shouldn't bring the whole process down.

teor2345 replied:
Ok, so I think I’m missing some context here.

What’s the impact on Subspace if each of these tasks fails?
- last-domain-execution-receipt-request-handler
- block-request-handler
- syncing

Let’s assume a failure on most or all of the nodes on the network, because that could easily happen in the logs and we wouldn’t know about it. (Because the node won’t go down, and most people don’t read logs.)

nazar-pc commented on crates/subspace-node/src/commands/run.rs:
```diff
@@ -153,6 +148,10 @@ pub async fn run(run_options: RunOptions) -> Result<(), Error> {
 
+            subspace_link = partial_components.other.subspace_link.clone();
+
+            segment_headers_store.replace(partial_components.other.segment_headers_store.clone());
```
`segment_headers_store` doesn't seem to be used anywhere, we discussed this before.

nazar-pc commented on crates/subspace-service/src/domains/snap_sync_orchestrator.rs:
```diff
@@ -0,0 +1,62 @@
+use tracing::debug;
+
+/// Synchronizes consensus and domain chain snap sync.
+pub struct SnapSyncOrchestrator {
```
I thought we agreed that this orchestrator is not needed, yet it still exists. It should be removed as unnecessary.

shamil-gadelshin replied:
We agreed that we need to provide only two shared variables an atomic bool for the sync oracle and a receiver for the target block. The rest was removed. The current orchestrator is just a container for those variables that are packed together most of the time. It is just more convenient to pass those variables in the same struct even if it is possible to remove it completely.

nazar-pc replied:
There is also a channel in it. I would have expected to remove this orchestrator completely, have some config data structure in domains if necessary and generally not have methods like `unblock_consensus_snap_sync` that need to be called in strategic moments.

teor2345 reviewed (dismissed):
Looks good to me. I also confirmed it’s essentially the same as your previous PR, here’s the compare view:
https://github.com/autonomys/subspace/compare/full-domain-snap-sync2...full-domain-snap-sync

Sorry to bring this up late in the process, but there’s no test code or test changes here. I understand that syncing code can be hard to test using unit tests. What’s the test plan for this feature?

Oh and it looks like there’s a minor error handling conflict, can you resolve that using a merge?

nazar-pc commented on domains/service/Cargo.toml:
```diff
@@ -63,8 +65,10 @@ sp-subspace-mmr = { version = "0.1.0", path = "../../crates/sp-subspace-mmr" }
 sp-transaction-pool = { git = "https://github.com/subspace/polkadot-sdk", rev = "5871818e1d736f1843eb9078f886290695165c42" }
 subspace-core-primitives = { version = "0.1.0", path = "../../crates/subspace-core-primitives" }
 subspace-runtime-primitives = { version = "0.1.0", path = "../../crates/subspace-runtime-primitives" }
+subspace-service = { version = "0.1.0", path = "../../crates/subspace-service" }
```
This is very problematic. You add a gigantic crate as a dependency to another gigantic crate. And the only purpose is to access `ConsensusChainSyncParams`, which most likely doesn't belong to `subspace_service` in the first place.

shamil-gadelshin replied:
I can get rid of this dependency by moving the struct. However, I would need to add `sc-consensus-subspace` dependency to `domain-operator` (block import streams dependency in the struct). Is it better?

nazar-pc replied:
Not really, domains should not rely on Subspace consensus code either

nazar-pc commented on crates/subspace-node/src/commands/run.rs:
```diff
@@ -307,17 +309,30 @@ pub async fn run(run_options: RunOptions) -> Result<(), Error> {
+                                    network_service: consensus_chain_network_service,
+                                    sync_service: consensus_chain_sync_service,
+                                    backend: consensus_chain_node.backend.clone(),
+                                    subspace_link,
```
`subspace_link` is a low-level-ish thing that should not be available to domains. If you need a specific stream then make it an argument of the domains and send it here. No need to send the whole `subspace_link` and

nazar-pc commented on domains/client/domain-operator/Cargo.toml:
```diff
@@ -39,6 +44,7 @@ sp-trie = { git = "https://github.com/subspace/polkadot-sdk", rev = "5871818e1d7
 sp-weights = { git = "https://github.com/subspace/polkadot-sdk", rev = "5871818e1d736f1843eb9078f886290695165c42" }
 subspace-core-primitives = { version = "0.1.0", path = "../../../crates/subspace-core-primitives" }
 subspace-runtime-primitives = { version = "0.1.0", path = "../../../crates/subspace-runtime-primitives" }
+subspace-service = { version = "0.1.0", default-features = false, path = "../../../crates/subspace-service" }
```
Ouch, this is way more problematic than the other one. This is a gigantic dependency added to a lower-level crate, which will significantly worsen compile times and will serialize compilation order even more.

shamil-gadelshin replied:
`domain-operator` hosts `domain snap sync feature` it requires both `mmr_sync` and `consensus snap sync` features from `subspace_service`. There are at least two ways to mitigate the existing dependency: either create a separate crate for sync features (both `subspace_service` and `domain_operator` will depend on it) or introduce several traits and wrapping structures to pass incapsulated functions to domains (`domain_operator` contains traits which are used by `subspace_node` to pass wrapping structures to domains). Either way, it seems more complicated than keeping an existing dependency. Do you see a simpler way or do you think the new complexity is justified from a dependency graph perspective?

nazar-pc replied:
I certainly expect a separate crate (or several) for sync, we'll have even more code once we upgrade our sync implementation to implement custom `SyncStrategy` after latest Substrate upgrade. But ideally operator wouldn't depend on consensus sync either, but rather on some shared crate with common primitives/utilities.

nazar-pc commented on crates/subspace-service/src/domains.rs:
```diff
@@ -1,229 +1,33 @@
 
-impl<Block, Client, NR> LastDomainBlockInfoReceiver<Block, Client, NR>
+/// Provides parameters for domain snap sync synchronization with the consensus chain snap sync.
+pub struct ConsensusChainSyncParams<Block, CNR>
```
Everything included in this struct is already public and can be accessed in `subspace-node`, this struct should not exist in `subspace-service`, it doesn't belong here.

shamil-gadelshin replied:
I answered above: the question remains about the dependency for block import notification stream.

nazar-pc replied:
Stream is just a stream. Domains already use some of the streams that are derived from consensus logic, yet they don't depend (I hope) on consensus crate for this.

nazar-pc reviewed (changes requested):
This will require more work to structure it in a more streamlined way.
There is still too much interleaving and coupling between components that should not know about each other existence.
From consensus knowing about domain sync to domain operator depending on the whole consensus chain service.

I see you have updated MMR stuff, that should have been one PR.
Then I see you want to reuse some sync implementation components from consensus in domains, the first step there would be to move those components into their own crate (long overdue, something like `sc-subspace-sync`, probably one PR) and then probably moving some generic stuff (like `sc-subspace-sync-common`, either second PR or together with the previous one).
I understand that simply making things public where they are and adding a dependency is quick and easy, but with this approach we will quickly get into recursive dependencies and there should be some reasonable hierarchy to dependencies.

After sync is a separate crate that doesn't depend on quite literally the whole subspace node implementation (minus CLI), this PR can be rebased, then we can get rid of `domain_snap_sync_finished`, `SnapSyncOrchestrator`, `ConsensusChainSyncParams` and I think it will be okay on consensus side.

teor2345 reviewed (commented):
Maybe these tasks should be essential?

shamil-gadelshin committed:
Extract syncing engine impl into separate crate `subspace-sync`

shamil-gadelshin committed:
Update sync oracle for domain snap sync.

shamil-gadelshin committed:
Remove MMR-sync from domain snap-sync

shamil-gadelshin committed:
Move snap-sync structures to domains.

teor2345 commented on domains/client/domain-operator/src/snap_sync.rs:
```diff
@@ -12,22 +13,107 @@ use sc_network_common::sync::message::{
 use tracing::{debug, error, trace, Instrument};
 
+/// Notification with number of the block that is about to be imported and acknowledgement sender
+/// that can be used to pause block production if desired.
```
As far as I remember, block production gets paused if the acknowledgements are missing or slow?
(This phrasing, and the phrasing below, make it sound like sending on the channel pauses production.)
```suggestion
/// that pauses block production until the previous block is acknowledged.
```

teor2345 commented on domains/client/domain-operator/src/snap_sync.rs:
```diff
@@ -12,22 +13,107 @@ use sc_network_common::sync::message::{
-use subspace_service::sync_from_dsn::snap_sync_engine::SnapSyncingEngine;
+use subspace_sync::snap_sync_engine::SnapSyncingEngine;
+use tokio::sync::broadcast;
+use tokio::sync::broadcast::{Receiver, Sender};
```
It’s a bit confusing to mix `futures::channel::mpsc` and `tokio::sync::broadcast` channels in the same file. But I understand that it’s probably because they have different behaviour when they get full.

Can you change `Sender` and `Receiver` into `broadcast::Sender` and `broadcast::Receiver` to make the code more readable?

teor2345 reviewed (commented):
Thanks, this looks a bit clearer.

shamil-gadelshin committed:
Refactor domain snap-sync

teor2345 reviewed (commented):
Thanks, looks good to me.

I’m happy to leave the essential task question to another ticket or PR, but we’ll need to resolve conflicts for this to merge.

nazar-pc commented on domains/client/domain-operator/src/lib.rs:
```diff
@@ -95,6 +95,9 @@ use sc_network_sync::service::network::NetworkServiceHandle;
 use sc_utils::mpsc::TracingUnboundedSender;
+pub use snap_sync::{
+    BlockImportingAcknowledgement, ConsensusChainSyncParams, SnapSyncOrchestrator,
+};
```
Can you simply make `snap_sync` module public instead of re-exporting?

nazar-pc commented on crates/subspace-service/Cargo.toml:
```diff
@@ -88,6 +88,7 @@ subspace-kzg = { version = "0.1.0", path = "../../shared/subspace-kzg" }
 subspace-networking = { version = "0.1.0", path = "../subspace-networking" }
 subspace-proof-of-space = { version = "0.1.0", path = "../subspace-proof-of-space" }
 subspace-runtime-primitives = { version = "0.1.0", path = "../subspace-runtime-primitives" }
+subspace-sync = { version = "0.1.0", path = "../../shared/subspace-sync" }
```
This commit should have been squashed with commit that introduced the crate. Otherwise you have a history disconnect due to file with implementation added there not linked to the file with implementation deleted here, Git will not preserve history of the file because of it, which is annoying.

nazar-pc commented on shared/subspace-sync/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "subspace-sync"
```
```suggestion
name = "sc-subspace-sync-common"
```

nazar-pc commented on domains/client/domain-operator/src/snap_sync.rs:
```diff
@@ -23,12 +23,12 @@ use std::sync::Arc;
 /// Notification with number of the block that is about to be imported and acknowledgement sender
-/// that can be used to pause block production if desired.
+/// that pauses block production until the previous block is acknowledged.
+
```
```suggestion
```

nazar-pc commented on crates/subspace-node/src/commands/run/domain.rs:
```diff
@@ -468,6 +477,13 @@ pub(super) async fn run_domain(
+        consensus_chain_sync_params
+            .as_ref()
+            .map(|params| params.snap_sync_orchestrator.domain_snap_sync_finished()),
+    ));
```
Still don't really understand why consensus syncing oracle needs to check if domain snap sync hash finished...

nazar-pc committed:
Rename sync oracle

shamil-gadelshin committed:
Merge branch 'main' into full-domain-snap-sync2

shamil-gadelshin committed:
Change MMR-sync target block.

shamil-gadelshin committed:
Rename subspace-sync -> sc-subspace-sync-common
